### PR TITLE
xdsl: syntactic sugar for Operand and OpResult DSL

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

File: docs/irdl.ipynb [hunk 1]
```diff
@@ -721,6 +721,8 @@
    "source": [
     "Similar to parameters for `ParametrizedAttribute`, operands and results are added using fields containing `OperandDef` and `ResultDef`, which each contain an attribute constraint. The order correspond to the operand and result order, and the constraint applies on the SSA variable type.\n",
     "\n",
+    "!!\n",
```
Comment (superlopuh) [inline docs/irdl.ipynb:724]: Need to write a new comment here

**Comment by superlopuh** [comment]
@georgebisbas @webmiche Which syntax do you prefer?

``` python
# Current PR syntax, should probably stick with either _/Var/Opt or _/list/|None
class Op:
  in0: Annotated[Operand, i32]
  in1: Annotated[VarOperand, i32]
  in2: Annotated[OptOperand, i32]

  out0: Annotated[OpResult, i32]
  out1: Annotated[list[OpResult], i32]
  out2: Annotated[OpResult | None, i32]
```

**Comment by webmiche** [comment]
AFAIU, the options are the upper or the lower version, right? I personally prefer the upper version. It hides the fact that these are just type hints a bit better and makes it feel more like an EDSL. Plus, I think the name `Operand` describes better what you are thinking of when designing an operation than `OpResult`, even though this is what it corresponds to.

**Comment by superlopuh** [comment]
I think I've made it more confusing than I could've by being lazy :) Here's what I meant:
``` Python
# Option A
class Op:
  in0: Annotated[Operand, i32]
  in1: Annotated[VarOperand, i32]
  in2: Annotated[OptOperand, i32]

  out0: Annotated[OpResult, i32]
  out1: Annotated[VarOpResult, i32]
  out2: Annotated[OptOpResult, i32]

# Option B
class Op:
  in0: Annotated[Operand, i32]
  in1: Annotated[list[Operand,] i32]
  in2: Annotated[Operand | None, i32]

  out0: Annotated[OpResult, i32]
  out1: Annotated[list[OpResult], i32]
  out2: Annotated[OpResult | None, i32]
```

It sounds like everyone prefers A so far, I'll refactor to match that

**Comment by webmiche** [comment]
Yes, I prefer that! Thanks for making it clear :)

**Comment by math-fehr** [comment]
Yeah, so I agree on having `Opt/Var` rather than `list/|None`!
For having `Result` instead of `OpResult`, I think that should be fine, since we do not have anything else that can be a "result". Also, I don't think that a `Result` type is used elsewhere (and if it is, we can also rename it during import).

**Comment by superlopuh** [comment]
The Result thing seems dangerous for some reason, my `Swift`/`Rust` reflexes are activated

**Review by georgebisbas — APPROVED** [review]

File: xdsl/irdl.py [hunk 2]
```diff
@@ -473,13 +488,33 @@ def from_pyrdl(pyrdl_def: type[_OpT]) -> OpDef:
                 continue
             args = get_args(field_type)
 
-            if isinstance(args[-1], OperandDef):
-                op_def.operands.append((field_name, args[-1]))
-            elif isinstance(args[-1], ResultDef):
-                op_def.results.append((field_name, args[-1]))
+            fail = False
+            if len(args) == 3:
```
Comment (math-fehr) [inline xdsl/irdl.py:492]: Actually, we shouldn't reason about the size of the arguments of `Annotated`.
For instance, `Annotated[A, Annotated[B, C]]` is simplified by Python to `Annotated[A, B, C]`.

**Review by math-fehr — APPROVED** [review]
Just added a comment on the discovery of parameters, otherwise once we merge it this should be merge asap.

**Comment by superlopuh** [comment]
I don't understand what it's unhappy about...

**Comment by math-fehr** [comment]
I'm confused as well.
It seems you have an error on a `Dealloca` operation, but this operation only exists in `main`, and is not rebased in this branch.
So yes, you should rebase it, but I don't understand why this triggers an error? I mean it's good that it does, but how does that work?
@webmiche, do you know why this happens?

**Comment by webmiche** [comment]
Honestly, I am confused as well. This happened after rebasing, right? And this is not something you can trigger locally?

I am honestly not sure how type hints work exactly. I can see that `Operand ist an `Annotaed[SSAValue, OperandDef]`, which might cause problems as OperandDef is not imported here. Though the fact that it is this operation failing is weird. I would have expected other tests to fail as well, if this actually is the issue...

And that a rebase triggers this behavior is also very weird. If my suggestion was the issue, this hsould have been broken before the rebase already...

**Comment by math-fehr** [comment]
So locally, everything works.
However, it seems that the error we get here, would be the error we get by rebasing the project, and then running the tests.
Is the CI always rebasing before running? That would be actually awesome, but I do not know if it is what's being done or not.

**Comment by webmiche** [comment]
Honestly, I cant imagine the CI to rebase before running. That would limit its use quite a bit for merge based projects...

Afaiu, you specify a branch/tag and this gets checked out. No rebasing and nothing...

**Comment by math-fehr** [comment]
So then I don't understand why we get an error here that is triggered in the `Dealloca` class, which is only there in `main`?
Do you think you could look into it?

**Comment by webmiche** [comment]
Ah wait I didn't realize that it was the DeallocA operation. You are right, it seems to be rebasing or potentially merging. I will look into this...

**Comment by webmiche** [comment]
Okay, I read some stuff, mainly a [stackoverflow](https://stackoverflow.com/questions/63594658/git-refs-merge-vs-head-in-pull-request) and [this](https://discourse.drone.io/t/github-claims-that-merge-refs-are-undocumented-feature/1100).

It seems to work as follows: Github runs the CI on a ref called `refs/pull/255/merge` (where 255 is the number of the current PR). This ref essentially contains the future state of the repo, so after merging the PR. This is to show what would happen, if this PR was merged and not just reflects what is actually there on the branch. You can see this ref on the CI btw, but it seems to be undocumented behavior, so we shouldn't rely on this actually working exactly like this.

So everything fine, github is smarter than we assumed I guess :)

**Comment by math-fehr** [comment]
Well, that's a win then!
The only thing we should be careful of then is when checked passed, but new commits were added after. In this case, GitHub won't check it again (because it could be too costly). So if check passes, it does not necessarily means that they will after a merge.

**Comment by math-fehr** [comment]
So after a rebase, we should merge it!

**Comment by math-fehr** [comment]
I made the rebase, is it good for you @webmiche, so we can merge?

**Comment by webmiche** [comment]
I am fine with merging, though I would prefer if we could add tests for the added lines that are not tested yet. AFAIU, that should be in the irdl file where several failing cases are not covered.

I am also fine with postponing this to a later PR, if you prefer to get this one merged asap.

**Comment by math-fehr** [comment]
I'll make a separate PR for it now!

**Pull request merged**